var/spack/repos/builtin/packages/py-smartsim/package.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+from spack.package import *
+
+
+class PySmartsim(PythonPackage):
+    """A workflow library to easily integrate machine learning libraries with
+    high performance computing simulations and applications
+    """
+
+    homepage = "https://www.craylabs.org/docs/overview.html"
+    git = "https://github.com/CrayLabs/SmartSim"
+    pypi = "smartsim/smartsim-0.4.2.tar.gz"
+
+    maintainers("MattToast")
+
+    version("0.5.0", sha256="35b36243dc84af62261a7f772bae92f0b3502faf01401423899cb2a48339858c")
+    version("0.4.2", sha256="ab632ff7d036e73822ddc5081fe85ea69c48d8be53ad0e6e487e9193eb3410f6")
+
+    variant("torch", default=True, description="Build with the pytorch backend")
+    variant("cuda", default=False, description="Use CUDA")
+    variant("rocm", default=False, description="Use ROCm")
+
+    depends_on("python@3.8:3.10", type=("build", "run"))
+    depends_on("py-wheel", type=("build",))
+    depends_on("py-setuptools", type=("build",))
+
+    depends_on("py-psutil@5.7.2:", type=("build", "run"))
+    depends_on("py-coloredlogs@10:", type=("build", "run"))
+    depends_on("py-tabulate@0.8.9:", type=("build", "run"))
+    depends_on("py-redis@4.5:", type=("build", "run"))
+    depends_on("py-tqdm@4.56:", type=("build", "run"))
+    depends_on("py-filelock@3.4:", type=("build", "run"))
+    depends_on("py-protobuf@3.20:3", type=("build", "run"))
+
+    # Companion libs
+    depends_on("py-smartredis@0.4.1", type=("build", "run"), when="@0.5.0")
+    depends_on("py-smartredis@0.4.0", type=("build", "run"), when="@0.4.2")
+
+    # Backends
+    depends_on("redis@7.0.5:", type=("build", "run"))
+
+    depends_on("redis-ai@1.2.7:", type=("build", "run"))
+    depends_on("redis-ai+cuda", type=("build", "run"), when="+cuda")
+    depends_on("redis-ai+rocm", type=("build", "run"), when="+rocm")
+
+    # ML Deps
+    with when("+torch"):
+        depends_on("redis-ai+torch", type=("build", "run"))
+        depends_on("py-torch@1.11:", type=("build", "run"))
+        depends_on("py-torch+cuda+cudnn", type=("build", "run"), when="+cuda")
+        depends_on("py-torch+rocm", type=("build", "run"), when="+rocm")
```

Comment:
I would add all of:
```python
        "1.2.7": {                                                                       
            "tensorflow": "2.8.0",                                                       
            "onnx": "1.11.0",                                                            
            "skl2onnx": "1.11.1",                                                        
            "onnxmltools": "1.11.1",                                                     
            "scikit-learn": "1.1.1",                                                     
            "torch": "1.11.0",                                                           
            "torch_cpu_suffix": "+cpu",                                                  
            "torch_cuda_suffix": "+cu113",                                               
            "torchvision": "0.12.0",                                                     
        },      
```

Reply:
I was actually hoping to break each of these out into their own variants to allow for finer grained control than SmartSim would naturally allow. These deps are only really needed if the corresponding ml backend for RedisAI is also installed for SmartSim's data loaders. e.g:

```py
with when("+torch"):
    depends_on("redis-ai+torch")
    depends_on("py-torch@...")
    ...

with when("+tensorflow"):
    depends_on("redis-ai+tensorflow")
    depends_on("py-tensorflow@...")
    ...

with when("+onnxruntime"):
    depends_on("redis-ai+onnxruntime")
    depends_on("py-onnx@...")
    depends_on("py-skl2onnx@...")
    ...
```
Currently only the Torch backed has been added for RedisAI spack package (thus it is the only variant currently implemented), but I was hoping to add more in follow up PRs as time allows.

Reply:
I'm fine with that since you're the one maintaining the recipe